Fw/Tlm/TlmPacket.cpp

```diff
@@ -10,91 +10,146 @@
 
 namespace Fw {
 
-    TlmPacket::TlmPacket() : m_id(0) {
+    TlmPacket::TlmPacket() : m_numEntries(0) {
         this->m_type = FW_PACKET_TELEM;
+        this->m_tlmBuffer.resetSer();
     }
 
     TlmPacket::~TlmPacket() {
     }
 
-    SerializeStatus TlmPacket::serialize(SerializeBufferBase& buffer) const {
-        SerializeStatus stat;
-#if !FW_AMPCS_COMPATIBLE
-        stat = serializeBase(buffer);
-        if (stat != FW_SERIALIZE_OK) {
+    SerializeStatus TlmPacket::resetPktSer() {
+        this->m_tlmBuffer.resetSer();
+        // reset packet count
+        this->m_numEntries = 0;
+        // serialize descriptor
+        return this->serializeBase(this->m_tlmBuffer);
+    }
+
+    SerializeStatus TlmPacket::resetPktDeser() {
+        this->m_tlmBuffer.resetDeser();
+        // deserialize descriptor
+        // The function deserializeBase inherited from ComPacket deserializes a
+        // value of type FwPacketDescriptorType from this->m_tlmBuffer and stores it
+        // into this->m_type.
+        Fw::SerializeStatus stat = this->deserializeBase(this->m_tlmBuffer);
+        if (stat != Fw::SerializeStatus::FW_SERIALIZE_OK) {
+            return stat;
+        }
+        // make sure that this->m_tlmBuffer stores a telemetry packet
+        if (this->m_type != FW_PACKET_TELEM) {
+            return Fw::SerializeStatus::FW_DESERIALIZE_TYPE_MISMATCH;
```

Comment:
Should we set this->m_type back to FW_PACKET_TELEM here? Otherwise we leave the component in an inconsistent state. Even better would be not to call deserializeBase at all, but do the deserialization into a local variable and check that against FW_PACKET_TELEM. I don't see why we should ever have to write a value to m_type after assigning it in the constructor -- it should always be FW_PACKET_TELEM.

Reply:
This is not a component, it is a helper class, so my expectation is that it would be created on the stack or somewhere temporarily for use in extracting channels. If we wanted to do anything, it would be to assert since the `deserializeBase()` should get it. I decided not to, since this could be a bad buffer passed in to the class. I thought it best to punt to the user rather than unilaterally assert or hack in the packet type. I'm not sure hacking in the packet type would accomplish anything since the data is probably bad anyway and the user should decide what to do with it. This is a little hand-wavy since we don't really have a use chase for deserializing on-board yet, so I was trying to implement a plausible way to get data out. I would leave the implementation as-is.

Reply:
OK, thanks. Let's leave as-is for now. It's still not clear to me why we need the m_type field in the base class. Maybe it's for run-time type identification? This is a broader F Prime issue though.